src/GafferScene/IECoreScenePreview/PrimitiveAlgo.cpp

```diff
@@ -0,0 +1,1188 @@
+		{
+			const MeshPrimitive *p = static_cast< const MeshPrimitive *>( primitives[i].first );
+
+			countCorners.push_back( p->cornerIds()->readable().size() );
+			accumCorners.push_back( totalAccumCorners );
+			totalAccumCorners += countCorners.back();
+
+			countCreases.push_back( p->creaseLengths()->readable().size() );
+			accumCreases.push_back( totalAccumCreases );
+			totalAccumCreases += countCreases.back();
+
+			countCreaseIds.push_back( p->creaseIds()->readable().size() );
+			accumCreaseIds.push_back( totalAccumCreaseIds );
+			totalAccumCreaseIds += countCreaseIds.back();
+		}
+
+		if( totalAccumCorners )
+		{
+			resultCornerIdsData = new IntVectorData;
+			resultCornerIdsData->writable().resize( totalAccumCorners );
+			resultCornerSharpnessesData = new FloatVectorData;
+			resultCornerSharpnessesData->writable().resize( totalAccumCorners );
+		}
+
+		if( totalAccumCreases )
+		{
+			resultCreaseLengthsData = new IntVectorData;
+			resultCreaseLengthsData->writable().resize( totalAccumCreases );
+			resultCreaseSharpnessesData = new FloatVectorData;
+			resultCreaseSharpnessesData->writable().resize( totalAccumCreases );
+			resultCreaseIdsData = new IntVectorData;
+			resultCreaseIdsData->writable().resize( totalAccumCreaseIds );
+		}
+
+		result = resultMesh;
+	}
+	else if( resultTypeId == IECoreScene::CurvesPrimitiveTypeId )
+	{
+		resultCurves = new IECoreScene::CurvesPrimitive();
+		result = resultCurves;
+
+		resultVerticesPerCurveData = new IntVectorData;
+		resultVerticesPerCurveData->writable().resize( totalInterpolation[ PrimitiveVariable::Uniform ] );
+	}
+	else
+	{
+		result = new IECoreScene::PointsPrimitive( totalInterpolation[ PrimitiveVariable::Vertex ] );
+	}
+
+	//
+	// Allocate storage for the primitives variables
+	//
+
+	for( auto &[name, varInfo] : varInfos )
+	{
+		if( varInfo.invalid )
+		{
+			continue;
+		}
+
+		varInfo.accumDataSizes.reserve( varInfo.numData.size() );
+		size_t accumDataSize = 0;
+		for( unsigned int i : varInfo.numData )
+		{
+			varInfo.accumDataSizes.push_back( accumDataSize );
+			accumDataSize += i;
+		}
+
+		PrimitiveVariable &p = result->variables.emplace(name, PrimitiveVariable() ).first->second;
+
+		p.data = IECore::runTimeCast<Data>( IECore::Object::create( varInfo.typeId ) );
+
+		IECore::setGeometricInterpretation( p.data.get(), varInfo.interpret );
+		p.interpolation = varInfo.interpolation;
+		Canceller::check( canceller );
+		dataResize( p.data.get(), accumDataSize );
+
+		if( varInfo.indexed )
+		{
+			p.indices = new IntVectorData();
+			Canceller::check( canceller );
+			p.indices->writable().resize( totalInterpolation[ varInfo.interpolation ] );
+		}
+	}
+
+	//
+	// Now a big parallel loop where we do the majority of actual work - copying all the primvar and topology
+	// data to the destination.
+	//
+
+	tbb::task_group_context taskGroupContext( tbb::task_group_context::isolated );
+
+	tbb::parallel_for(
+		tbb::blocked_range<size_t>( 0, primitives.size() ),
```

Comment:
Depends entirely on the workload - merging many small objects is a plausible workload that would benefit a lot from this. But merging a few enormous objects is also plausible, and wouldn't benefit from this. It would be relatively trivial to break up the copying of data across threads to speed up the latter case ... though handling the indices would be a bit more annoying. Technically pretty trivial to at least break up different primvars and topology across different threads, so I'm not immediately sure how to express that cleanly.